### PR TITLE
Various parser improvements

### PR DESCRIPTION
**Pull request by acharis**
since pushing, i've experimented more with mysql and i think i need to tweak this some.
not and ! bind differently:
```
 mysql> select !1 + 1;  // 1
 mysql> select not 1 + 1; // 0
```

(some changes from running gofmt)

### CODE REVIEW

File: go/vt/sqlparser/token.go [hunk 1]
```diff
@@ -403,11 +403,13 @@ func (tkn *Tokenizer) Scan() (int, []byte) {
 				return int(ch), nil
 			}
 		case '!':
-			if tkn.lastChar == '=' {
+			switch tkn.lastChar {
```
Comment (bbeaudreault) [inline go/vt/sqlparser/token.go]: can this be simplified not use the switch, since it's basically the same as what it was but with a different return statement?

File: go/vt/sqlparser/parse_test.go [hunk 2]
```diff
@@ -307,7 +310,9 @@ func TestValid(t *testing.T) {
 	}, {
 		input: "select /* / */ 1 from t where a = b / c",
 	}, {
-		input: "select /* % */ 1 from t where a = b % c",
```
Comment (bbeaudreault) [inline go/vt/sqlparser/parse_test.go]: did we lose a test here?

**Comment by sougou** [comment]
@acharis You have to specify a precedence for `!` here: https://github.com/youtube/vitess/blob/master/go/vt/sqlparser/sql.y#L85, as defined by https://dev.mysql.com/doc/refman/5.7/en/operator-precedence.html. Note that the order is reversed for yacc.

**Comment by acharis** [comment]
@sougou please see https://github.com/HubSpot/vitess/blob/not-exclamation/go/vt/sqlparser/sql.y#L907  which is current rev

**Comment by acharis** [comment]
@sougou and thanks for taking a look. i'm still working on this, so i'll ping you when it's ready for review. i'm going to add some more features in over the next few days to cut down on PRs

File: go/vt/sqlparser/ast.go [hunk 3]
```diff
@@ -1010,6 +1019,7 @@ const (
 	NotRegexpStr         = "not regexp"
 	JSONExtractOp        = "->"
 	JSONUnquoteExtractOp = "->>"
+	CollateStr           = "collate"
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: I think this was removed in master and its existence here is a result of merge conflict

File: go/vt/sqlparser/ast.go [hunk 4]
```diff
@@ -1465,7 +1479,7 @@ func (node *IntervalExpr) WalkSubtree(visit Visit) error {
 
 // CollateExpr represents dynamic collate operator.
 type CollateExpr struct {
-	Expr Expr
+	Expr    Expr
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: whitespace

File: go/vt/sqlparser/ast.go [hunk 5]
```diff
@@ -1570,6 +1584,110 @@ func (node *ValuesFuncExpr) WalkSubtree(visit Visit) error {
...
+	Type *ConvertType
+}
+
+// Format formats the node.
+func (node *ConvertExpr) Format(buf *TrackedBuffer) {
+	buf.Myprintf("convert( ")
+	if node.Expr != nil {
+		buf.Myprintf("%v", node.Expr)
+	}
+	buf.Myprintf(" , ")
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: should this line be in the above if statement? Otherwise if Expr is nil, we will see something like `convert( , node.Type)` or `convert( ,)` if Type is also nil.

Also, is it possible in the grammar for these to be nil?

File: go/vt/sqlparser/sql.y [hunk 6]
```diff
@@ -950,10 +985,157 @@ value_expression:
...
+   $$ = NaturalLanguageModeWithQueryExpansionStr
+ }
+| WITH QUERY EXPANSION
+ {
+   $$ = QueryExpansionStr
+ }
+
+
+convert_type:
+BINARY openb INTEGRAL closeb
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: I think the spacing is off on this entire `convert_type` grammar

File: go/vt/sqlparser/sql.y [hunk 7]
```diff
@@ -950,10 +985,157 @@ value_expression:
...
+| DATETIME
+{
+	$$ = &ConvertType{Type: Datetime}
+}
+| JSON
+{
+	$$ = &ConvertType{Type: Json}
+}
+
+opt_integer_keyword:
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: should this be `integer_keyword_opt` to stick with convention?

File: go/vt/sqlparser/sql.y [hunk 8]
```diff
@@ -950,10 +985,157 @@ value_expression:
...
+/*empty*/
+| INTEGER
+
+charset_info:
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: spacing in this one too

File: go/vt/sqlparser/sql.y [hunk 9]
```diff
@@ -950,10 +985,157 @@ value_expression:
...
+| CAST openb expression AS convert_type closeb
+  {
+    $$ = &ConvertExpr{Expr: $3, Type: $5}
+  }
+| MATCH openb column_list closeb AGAINST openb value_expression match_option closeb
+{
+  $$ = &MatchExpr{Columns: $3, Expr: $7, Option: $8}
+}
+
+match_option:
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: should probably be `match_type_opt` or something, to go with convention

File: go/vt/sqlparser/ast.go (hunk 3 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: yes, i saw that during merge. i'll remove.

File: go/vt/sqlparser/ast.go (hunk 4 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: gofmt

File: go/vt/sqlparser/ast.go (hunk 5 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: fixed.

File: go/vt/sqlparser/ast.go (hunk 4 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: Oh right, it's aligning it with Charset below.  👍

File: go/vt/sqlparser/sql.y (hunk 9 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: opt means optional (and should be prefix, but we're doing suffix)

though this option is, in fact, optional, i thought opt_match_option was a bit too much

Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: `search_modifier_opt` -- not too bad, that's what they call it in the mysql docs.

File: go/vt/sqlparser/sql.y (hunk 6 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: too true. fixed

File: go/vt/sqlparser/sql.y (hunk 7 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: fixed

File: go/vt/sqlparser/sql.y (hunk 8 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: done

File: go/vt/sqlparser/ast.go [hunk 10]
```diff
@@ -1570,6 +1583,102 @@ func (node *ValuesFuncExpr) WalkSubtree(visit Visit) error {
...
+	Binary   = "binary"
+	Char     = "char"
+	NChar    = "nchar"
+	Decimal  = "decimal"
+	Signed   = "signed"
+	Unsigned = "unsigned"
+	Date     = "date"
+	Datetime = "datetime"
+	Time     = "time"
+	Json     = "json"
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: golint tells me this var name should be `JSON`

go/vt/sqlparser/ast.go:1591:2: const Json should be JSON

Comment (acharis) [inline go/vt/sqlparser/ast.go]: and if you try that Json -> JSON you then get compiler errors.
i felt JsonType (and then, for consistency, DateType, BinaryType, etc) was too wordy.

Comment (acharis) [inline go/vt/sqlparser/ast.go]: but apparently i didn't feel NaturalLanguageModeWithQueryExpansionStr was too wordy

**Comment by acharis** [comment]
ok, i just added support for the BINARY operator to this PR

File: go/vt/sqlparser/ast.go [hunk 11]
```diff
@@ -1121,6 +1129,29 @@ func (node *ExistsExpr) WalkSubtree(visit Visit) error {
 	)
 }
 
+// ValExpr represents a value expression.
+type ValExpr interface {
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: i think this was accidentally included in a merge conflict

File: go/vt/sqlparser/ast.go [hunk 12]
```diff
@@ -1749,8 +1879,8 @@ func (node UpdateExprs) WalkSubtree(visit Visit) error {
 
 // UpdateExpr represents an update expression.
 type UpdateExpr struct {
-	Name ColIdent
-	Expr Expr
+	Name *ColName
+	Expr ValExpr
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: Merge conflict: This should be Expr

File: go/vt/sqlparser/ast.go [hunk 13]
```diff
@@ -148,10 +149,16 @@ const (
...
+	SQLCacheStr   = "SQL_CACHE "
+	SQLNoCacheStr = "SQL_NO_CACHE "
+)
+
 // Format formats the node.
 func (node *Select) Format(buf *TrackedBuffer) {
-	buf.Myprintf("select %v%s%s%v from %v%v%v%v%v%v%s",
-		node.Comments, node.Distinct, node.Hints, node.SelectExprs,
+	buf.Myprintf("select %s%v%s%s%v from %v%v%v%v%v%v%s",
+		node.Cache, node.Comments, node.Distinct, node.Hints, node.SelectExprs,
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: should this go after the comments, since it goes after the comments in the sql.y?

File: go/vt/sqlparser/sql.y [hunk 14]
```diff
@@ -938,10 +978,157 @@ value_expression:
   {
     $$ = $1
   }
+
 | VALUES openb sql_id closeb
   {
     $$ = &ValuesFuncExpr{Name: $3}
   }
+| CONVERT openb expression ',' convert_type closeb
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: I feel like this and below should be a value_expression, as all other things in value_expression reference

File: go/vt/sqlparser/sql.y [hunk 15]
```diff
@@ -938,10 +978,157 @@ value_expression:
...
+| INTEGER
+
+charset_info:
+  CHARACTER SET charset
+  {
+    $$ = $3
+  }
+| ASCII
+  {
+    $$ = "latin1"
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: are these supposed to be hardcoded like this?

File: go/vt/sqlparser/ast.go (hunk 13 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: yep

File: go/vt/sqlparser/sql.y (hunk 15 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: in mysql, they are (if you drill down)

File: go/vt/sqlparser/ast.go (hunk 12 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: fixed

File: go/vt/sqlparser/ast.go (hunk 11 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: fixed, thx for the catch

File: go/vt/sqlparser/sql.y (hunk 14 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: that would be more restrictive than mysql

Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: Ok, does it make sense to move up a level then?  Perhaps these should be in the `expression:` grammar itself? I don't have a huge objection here, just commenting that all of the recursive rules in `value_expression` grammar reference another `value_expression`, while these new ones reference `expression`

File: go/vt/sqlparser/sql.y (hunk 15 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: I think simplification would be better, mysql can handle the setting of real values, we mostly care about accurately re-representing the query on rewrite

File: go/vt/sqlparser/sql.y [hunk 16]
```diff
@@ -226,7 +241,7 @@ insert_statement:
     cols := make(Columns, 0, len($6))
     vals := make(ValTuple, 0, len($7))
     for _, updateList := range $6 {
-      cols = append(cols, updateList.Name)
+      cols = append(cols, updateList.Name.Name)
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: I think `Columns` should instead potentially be updated to be a `[]ColName`

File: go/vt/sqlparser/ast.go [hunk 17]
```diff
@@ -1504,6 +1516,31 @@ func (node *FuncExpr) IsAggregate() bool {
...
+type GroupConcatExpr struct {
+	Distinct  string
+	Exprs     SelectExprs
+	OrderBy   OrderBy
+	Separator string
+}
+
+// Format formats the node
+func (node *GroupConcatExpr) Format(buf *TrackedBuffer) {
+	buf.Myprintf("group_concat(%s %v %v separator %s)", node.Distinct, node.Exprs, node.OrderBy, node.Separator)
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: I think most of these fields are optional in the grammar: https://dev.mysql.com/doc/refman/5.7/en/group-by-functions.html#function_group-concat

File: go/vt/sqlparser/sql.y [hunk 18]
```diff
@@ -942,6 +1001,152 @@ value_expression:
   {
     $$ = &ValuesFuncExpr{Name: $3}
   }
+| CONVERT openb expression ',' convert_type closeb
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: Can you also support `CONVERT(foo USING bar)`? In terms of the second argument in that syntax, you can probably just use the existing charset grammar (i.q. sql_id) https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_convert. We don't really care about the actual contents of the arguments here except to properly re-represent them in the Format function of the ast.

File: go/vt/sqlparser/sql.y [hunk 19]
```diff
@@ -174,18 +189,23 @@ func forceEOF(yylex interface{}) {
...
 %%
 
 any_command:
-  command
+  command opt_semicolon
   {
     setParseTree(yylex, $1)
   }
 
+opt_semicolon:
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: semicolon_opt

File: go/vt/sqlparser/sql.y [hunk 20]
```diff
@@ -621,6 +654,16 @@ where_expression_opt:
     $$ = $2
   }
 
+//expression_or_default:
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: commented out

File: go/vt/sqlparser/sql.y [hunk 21]
```diff
@@ -876,6 +923,10 @@ value_expression:
   {
     $$ = &CollateExpr{Expr: $1, Charset: $3}
   }
+| BINARY value_expression %prec UNARY
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: I'd add tests to see how this plays with other unaries. Not sure how safe it is to overload that, but if it works than great.

File: go/vt/sqlparser/sql.y [hunk 22]
```diff
@@ -942,6 +1001,152 @@ value_expression:
...
+ {
+    $$ = NaturalLanguageModeWithQueryExpansionStr
+ }
+| WITH QUERY EXPANSION
+ {
+    $$ = QueryExpansionStr
+ }
+
+
+convert_type:
```
Comment (bbeaudreault) [inline go/vt/sqlparser/sql.y]: Should we simplify some of the below grammar? I'm not sure we need the validation in the actions, for instance. As with the other stuff, we're really just re-printing this stuff in Format. So I'm not sure what maintaining the truly proper syntax for each type gets us. If someone passes in the wrong arguments, it'll error out in mysql and propagate up.

File: go/vt/sqlparser/parse_test.go [hunk 23]
```diff
@@ -667,6 +679,17 @@ func TestValid(t *testing.T) {
...
+		input:  "select /* drop trailing semicolon */ 1 from dual;",
+		output: "select /* drop trailing semicolon */ 1 from dual",
+	}, {
+		input: "select /* cache directive */ SQL_NO_CACHE 'foo' from t",
+	}, {
+		input: "select binary 'a' = 'A' from t",
+	}, {
+		input: "select match(a) against ('foo') from t",
+	}, {
+		input: "select match(a1, a2) against ('foo' in natural language mode with query expansion) from t",
```
Comment (bbeaudreault) [inline go/vt/sqlparser/parse_test.go]: can you also add a test for the semi-colon? not seeing it here. anything else we're missing?

Comment (bbeaudreault) [inline go/vt/sqlparser/parse_test.go]: group_concat

File: go/vt/sqlparser/ast.go [hunk 24]
```diff
@@ -1373,24 +1383,30 @@ func (node *BinaryExpr) WalkSubtree(visit Visit) error {
...
 // Format formats the node.
 func (node *UnaryExpr) Format(buf *TrackedBuffer) {
 	if _, unary := node.Expr.(*UnaryExpr); unary {
-		buf.Myprintf("%c %v", node.Operator, node.Expr)
+		buf.Myprintf("%s %v", node.Operator, node.Expr)
 		return
 	}
-	buf.Myprintf("%c%v", node.Operator, node.Expr)
+	if node.Operator == BinaryStr {
+		buf.Myprintf("%s %v", node.Operator, node.Expr)
```
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: should we just make `BinaryStr = "binary "` to avoid the comparison?

File: go/vt/sqlparser/parse_test.go (hunk 23 above)
Comment (acharis) [inline go/vt/sqlparser/parse_test.go]: semicolon test is there (search for semicolon instead of semi-colon
group_concat test is coming

File: go/vt/sqlparser/ast.go [hunk 25]
```diff
@@ -1749,7 +1894,7 @@ func (node UpdateExprs) WalkSubtree(visit Visit) error {
 
 // UpdateExpr represents an update expression.
 type UpdateExpr struct {
-	Name ColIdent
+	Name *ColName
```
Comment (acharis) [inline go/vt/sqlparser/ast.go]: yikes, i'll look harder at that in a moment

File: go/vt/sqlparser/ast.go [hunk 26]
```diff
@@ -1534,6 +1575,110 @@ func (node *ValuesFuncExpr) WalkSubtree(visit Visit) error {
...
+	Type     string
+	Length   *SQLVal
+	Scale    *SQLVal
+	Operator string
+	Charset  string
+}
+
+// ConvertType.Type
+const (
+	BinaryType   = "binary"
```
Comment (acharis) [inline go/vt/sqlparser/ast.go]: correct, removed

File: go/vt/sqlparser/ast.go [hunk 27]
```diff
@@ -1504,6 +1520,31 @@ func (node *FuncExpr) IsAggregate() bool {
...
+type GroupConcatExpr struct {
+	Distinct  string
+	Exprs     SelectExprs
+	OrderBy   OrderBy
+	Separator string
+}
+
+// Format formats the node
+func (node *GroupConcatExpr) Format(buf *TrackedBuffer) {
+	buf.Myprintf("group_concat(%s%v%vseparator %s)", node.Distinct, node.Exprs, node.OrderBy, node.Separator)
```
Comment (acharis) [inline go/vt/sqlparser/ast.go]: nope

File: go/vt/sqlparser/ast.go (hunk 24 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: as it turns out, no.
if you, because you're silly, write "binary binary 'foo'" then to get the spacing right on that, as well as the normal case, we've got to do it this way.
if we make it "binary ", then the others have to be "+ " etc, and it looks wrong

File: go/vt/sqlparser/sql.y (hunk 22 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: done, looks better too

File: go/vt/sqlparser/sql.y (hunk 21 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: done

File: go/vt/sqlparser/sql.y (hunk 16 above)
Comment (acharis) [inline go/vt/sqlparser/sql.y]: rabbit-hole, that's another PR

File: go/vt/vtgate/planbuilder/dml.go [hunk 28]
```diff
@@ -27,8 +27,12 @@ func buildUpdatePlan(upd *sqlparser.Update, vschema VSchema) (*engine.Route, err
 	route := &engine.Route{
 		Query: generateQuery(upd),
 	}
+	updateTable, ok := upd.Table.Expr.(*sqlparser.TableName)
```
Comment (bbeaudreault) [inline go/vt/vtgate/planbuilder/dml.go]: Add a test for this if you haven't. I think there's an unsupported_cases.txt

File: go/vt/sqlparser/ast.go (hunk 25 above)
Comment (acharis) [inline go/vt/sqlparser/ast.go]: then we lose the table qualifier in update

File: go/vt/sqlparser/ast.go (hunk 24 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: this is pretty nit-picky, but honestly i think the extra space might be worth it to simplify the code.  an extra space in the case of `binary  <unary> 'foo'` doesn't really do anything.

File: go/vt/sqlparser/parse_test.go (hunk 23 above)
Comment (bbeaudreault) [inline go/vt/sqlparser/parse_test.go]: Nice, found it. 👍

**Comment by acharis** [comment]
@sougou I think I've just about wrapped up dev on this branch. Other things I'll work on belong in another PR. If you could take a look, that would be appreciated. The travis failure seems unrelated, and I can't repro locally.
